include/sys/dmu_objset.h

```diff
@@ -125,6 +126,12 @@ struct objset {
 	kmutex_t os_user_ptr_lock;
 	void *os_user_ptr;
 	sa_os_t *os_sa;
+
+	/* kernel thread to upgrade this dataset */
+	kthread_t *os_upgrade_thread;
+	kmutex_t os_upgrade_lock;
+	kcondvar_t os_upgrade_cv;
+	boolean_t os_upgrade_exit;
```

Comment:
It's a real case, we definitely have users with 1000+ datasets which all get mounted.  I'm aware of one user with 10,000+.

Starting them in `dmu_objset_own()` own might work.  I was giving this some thought too, and I think moving to a taskq in the spa could be a clean way to handle this.  Here's what I'm thinking:

* In `spa_activate()` you can create a new `spa->spa_upgrade_taskq` called `z_upgrade` to handle the upgrade.  It can be created it with 1 thread per-cpu to get some parallelism, and destroyed in `spa_deactivate()`.  We may want to get slightly fancier and create/destroy the taskq as needed so the taskq and it's worker threads don't hang around when they're not needed.  I like how you set this up to use a callback so this can be generic infrastructure, we may need to make use of this for other new features.

* In `dmu_objset_own()` you need only check if the dataset needs to be upgraded, use `taskq_dispatch()` to queue the upgrade function, and then store away the unique taskqid_t as say `os->os_upgrade_id`.  Alternately, it may be simpler to use a prealloced tqent if we can live with the limitation of one outstanding upgrade per dataset.

* Then in `dmu_objset_disown()` you'll want to set some flag to signal the taskq func to exit if executing and call `taskq_cancel_id()` when `os->os_upgrade_id != -1`.  After it completes you'll be guaranteed that the taskqid was removed from both the pending queue and is no longer running on an active worker thread.

What do you think?